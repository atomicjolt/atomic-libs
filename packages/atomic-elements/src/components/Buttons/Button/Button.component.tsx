import { forwardRef } from "react";
import { AriaButtonOptions } from "@react-aria/button";
import { mergeProps, useObjectRef } from "@react-aria/utils";

import { SpinnerLoader } from "../../Loaders/SpinnerLoader";
import {
  ExtendedSize,
  HasVariant,
  LoadingProps,
  RenderBaseProps,
  SuggestStrings,
} from "../../../types";
import { useFocusRing } from "@hooks/useFocusRing";
import { useRenderProps } from "@hooks/useRenderProps";
import { useButtonLink } from "@hooks/useButtonLink";
import { useContextProps } from "@hooks/useContextProps";
import { StyledButton } from "./Button.styles";
import { ButtonContext } from "./Button.context";
import { SlotProps } from "@hooks/useSlottedContext";

export type ButtonVariants = SuggestStrings<
  | "primary"
  | "secondary"
  | "link"
  | "success"
  | "error"
  | "inverted"
  | "content"
  | "border"
  | "ghost"
>;

interface ButtonRenderProps {
  isLoading: boolean;
  isPressed: boolean;
  isFocusVisible: boolean;
  isFocused: boolean;
}

export interface ButtonProps
  extends AriaButtonOptions<"button">,
    LoadingProps,
    RenderBaseProps<ButtonRenderProps>,
    HasVariant<ButtonVariants>,
    SlotProps {
  as?: "button" | "a";
  size?: ExtendedSize;
}

/** A button component that can be used to trigger actions or events
 *
 * @example <Button onPress={() => alert("Hello, world!")}>Click me</Button>
 */
export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(props, forwardedRef) {
    [props, forwardedRef] = useContextProps(ButtonContext, props, forwardedRef);

    const ref = useObjectRef(forwardedRef);

    const {
      isLoading = false,
      loadingLabel,
      loadingComplete = false,
      as = props.href ? "a" : "button",
      variant = "primary",
      size = "auto",
    } = props;

    const { buttonProps, isPressed } = useButtonLink(
      {
        ...props,
        elementType: as,
        "aria-label": isLoading ? loadingLabel : props["aria-label"],
      },
      ref
    );

    const { focusProps, isFocusVisible, isFocused } = useFocusRing();

    const renderProps = useRenderProps({
      componentClassName: "aje-btn",
      ...props,
      variant,
      size,
      values: { isPressed, isLoading, isFocusVisible, isFocused },
      selectors: {
        "data-pressed": isPressed,
        "data-loading": isLoading,
      },
    });

    return (
      <StyledButton
        as={as}
        ref={ref}
        {...mergeProps(buttonProps, focusProps, renderProps)}
      >
        {isLoading && (
          <SpinnerLoader
            isLoading={!loadingComplete}
            placement="absolute center"
          />
        )}
        {renderProps.children}
      </StyledButton>
    );
  }
);
