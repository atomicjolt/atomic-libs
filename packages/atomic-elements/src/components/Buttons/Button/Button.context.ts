import { ButtonProps } from ".";
import { createComponentContext } from "@utils/index";

export const ButtonContext = createComponentContext<ButtonProps>();
