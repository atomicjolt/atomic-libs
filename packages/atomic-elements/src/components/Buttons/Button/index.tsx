export { Button } from "./Button.component";
export type { ButtonProps, ButtonVariants } from "./Button.component";
