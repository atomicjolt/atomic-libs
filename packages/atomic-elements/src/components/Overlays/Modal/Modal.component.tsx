import React, { useContext, useRef } from "react";
import { AriaModalOverlayProps, useModalOverlay } from "@react-aria/overlays";
import { OverlayTriggerState, useOverlayTriggerState } from "react-stately";

import { useContextProps } from "@hooks/useContextProps";
import { useRenderProps } from "@hooks/useRenderProps";
import { OverlayTriggerStateContext } from "../OverlayTrigger/context";

import { BaseModalProps, ModalChildren } from "./Modal.types";
import { ModalContext } from "./Modal.context";
import { ModalWrapper } from "./Modal.styles";
import { ModalOverlay } from "./components/ModalOverlay";
import { ModalHeader } from "./components/ModalHeader";
import { ModalTitle } from "./components/ModalTitle";
import { ModalBody } from "./components/ModalBody";
import { ModalFooter } from "./components/ModalFooter";
import { useManageModalScroll } from "./hooks/useManageModalScroll";

export interface ModalProps extends BaseModalProps {
  children?: ModalChildren;
}

/**
 * Modal Component to render content overlayed on top of the page content.
 */
export function Modal(props: ModalProps) {
  let ref = useRef(null);
  [props, ref] = useContextProps(ModalContext, props, ref);

  const { children, centered = false, isCentered = centered, ...rest } = props;

  const contextState = useContext(OverlayTriggerStateContext);
  const localState = useOverlayTriggerState(props);
  const state = contextState ?? localState;

  if (!state.isOpen) {
    return null;
  }

  return (
    <ModalInternal state={state} isCentered={isCentered} {...rest}>
      {children}
    </ModalInternal>
  );
}

interface ModalInternalProps extends BaseModalProps, AriaModalOverlayProps {
  state: OverlayTriggerState;
  children: ModalChildren;
  triggerRef?: React.RefObject<HTMLElement | null>;
}

function ModalInternal(props: ModalInternalProps) {
  const {
    children,
    className,
    style,
    state,
    isCentered,
    isOpen,
    triggerRef,
    ensureVisible,
    ...rest
  } = props;
  const ref = useRef<HTMLDivElement | null>(null);
  const { modalProps, underlayProps } = useModalOverlay(props, state, ref);

  useManageModalScroll(
    {
      ensureVisible,
      modalRef: ref,
      scrollOptions: {
        behavior: "instant",
        block: "end",
        inline: "nearest",
      },
    },
    state
  );

  const renderProps = useRenderProps({
    componentClassName: "aje-modal",
    children,
    className,
    style,
  });

  return (
    <ModalOverlay isCentered={isCentered} {...underlayProps}>
      <ModalWrapper
        onClick={(e: React.MouseEvent<HTMLDivElement>) => e.stopPropagation()}
        ref={ref}
        data-placement={isCentered ? "center" : "top"}
        {...rest}
        {...modalProps}
        {...renderProps}
      >
        {typeof children === "function" ? children(state.close) : children}
      </ModalWrapper>
    </ModalOverlay>
  );
}

/** Wrapper around Flex to provide some sensible defaults for Modals */
Modal.Header = ModalHeader;

/** Wrapper around Heading to provide some sensible defaults for Modals */
Modal.Title = ModalTitle;

/** Wrapper around View to provide some sensible defaults for Modals */
Modal.Body = ModalBody;

/** Wrapper around Flex to provide some sensible defaults for Modals */
Modal.Footer = ModalFooter;
