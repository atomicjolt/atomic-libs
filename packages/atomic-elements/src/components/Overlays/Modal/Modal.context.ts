import React from "react";
import { ModalProps } from ".";
import { createComponentContext } from '@utils/index';

export const ModalContext = createComponentContext<ModalProps>();
