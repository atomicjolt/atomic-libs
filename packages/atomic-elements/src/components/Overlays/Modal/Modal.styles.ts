import styled from "styled-components";
import mixins from "../../../styles/mixins";
import { CheckboxWrapper } from "../../Inputs/Checkbox/Checkbox.styles";
import { ChooseLabel } from "../../Inputs/Inputs.styles";

export const ModalHeader = styled.div.attrs({ className: "aje-modal__top" })`
  display: flex;
  align-items: center;
  gap: 8px;
`;

export const ModalTitle = styled.h2.attrs({ className: "aje-modal__title" })<{
  fill?: boolean;
}>`
  ${mixins.Bold}
  margin: 0;
  font-size: 2rem;
  ${({ fill = true }) => fill && "flex: 1;"}
  line-height: 1;
  color: var(--text-clr);
`;

export const ModalBody = styled.div.attrs({ className: "aje-modal__main" })`
  ${mixins.Regular}
  padding: 16px 0 24px;
  font-size: 1.6rem;
  color: var(--text-clr);
  line-height: 1.5;
`;

export const ModalFooter = styled.div.attrs({ className: "aje-modal__bottom" })`
  display: flex;
  align-items: center;
  justify-content: flex-end;
  gap: 8px;
`;

export const ModalBackground = styled.div`
  position: fixed;
  min-height: 300px;
  inset: 0;
  background-color: hsla(0, 0%, 100%, 0.75);
  overflow: auto;
  display: flex;
  justify-content: center;
  align-items: flex-start;
  z-index: 1000;

  &.is-centered {
    display: grid;
    place-items: center;
    justify-content: initial;

    .aje-modal {
      margin: 0;
    }
  }
`;

export const ModalWrapper = styled.div`
  width: 90%;
  max-width: 800px;
  border-radius: 10px;
  background-color: var(--neutral50);
  padding: 24px;
  box-shadow: 0 10px 40px hsla(0, 0%, 0%, 0.25);
  margin-top: 56px;

  &.aje-modal--popup {
    max-width: 400px;
  }

  &.aje-modal--error {
    max-width: 400px;

    .aje-modal__top > i {
      color: var(--error700);
    }
  }

  &.aje-modal--developer-error {
    max-width: 1200px;
    background-color: #181818;

    ${ModalHeader} {
      color: var(--error700) !important;
    }

    ${ModalTitle} {
      color: var(--error700) !important;
    }
  }
`;
