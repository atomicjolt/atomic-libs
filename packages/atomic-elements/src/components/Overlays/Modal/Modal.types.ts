import { OverlayTriggerProps } from "react-stately";
import { ElementWrapper, RenderStyleProps } from "../../../types";
import { LayoutProps } from "@styles/layout";

export type ModalChildren =
  | React.ReactNode
  | ((close: () => void) => React.ReactNode);

export interface BaseModalProps
  extends RenderStyleProps<never>,
    OverlayTriggerProps,
    ElementWrapper<HTMLDivElement>,
    LayoutProps {
  /** Centers the modal within the viewport
   * @deprecated Use `isCentered` instead
   */
  centered?: boolean;

  /** Centers the modal within the viewport */
  isCentered?: boolean;

  /** Ensures the modal is visible by scrolling it into view.
   *
   * This is typically unnecessary since modals are automatically
   * centered in the browser viewport. However, when rendering
   * within an iframe that exceeds the parent viewport height,
   * the optimal placement cannot be determined. In this case,
   * the modal is positioned at the top of the page and
   * automatically scrolled into view.
   */
  ensureVisible?: boolean;
}
