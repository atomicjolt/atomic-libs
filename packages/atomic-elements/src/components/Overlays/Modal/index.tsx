export { Modal } from "./Modal.component";
export { ModalContext } from "./Modal.context";
export type { ModalProps } from "./Modal.component";
export type { BaseModalProps } from "./Modal.types";
