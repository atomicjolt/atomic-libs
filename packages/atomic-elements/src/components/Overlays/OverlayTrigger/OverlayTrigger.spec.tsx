import { act, render } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";
import { OverlayTrigger } from ".";
import { Button } from "../../Buttons/Button";
import { Modal } from "../Modal";

describe("OverlayTrigger", () => {
  it("Matches the closed snapshot", () => {
    const res = render(
      <OverlayTrigger>
        <Button>Trigger</Button>
        <Modal>
          <div>Content</div>
        </Modal>
      </OverlayTrigger>
    );
    expect(res.asFragment()).toMatchSnapshot();
  });

  it("Matches the open snapshot", () => {
    const onOpenChange = vi.fn();
    const res = render(
      <OverlayTrigger onOpenChange={onOpenChange}>
        <Button>Trigger</Button>
        <Modal>
          <div>Content</div>
        </Modal>
      </OverlayTrigger>
    );

    act(() => {
      res.getByText("Trigger").click();
    });

    expect(res.asFragment()).toMatchSnapshot();
    expect(onOpenChange).toHaveBeenCalledOnce();
    expect(onOpenChange).toHaveBeenCalledWith(true);
  });
});
