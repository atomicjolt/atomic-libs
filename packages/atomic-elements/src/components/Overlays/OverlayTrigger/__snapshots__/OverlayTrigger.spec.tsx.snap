// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OverlayTrigger > Matches the closed snapshot 1`] = `
<DocumentFragment>
  <button
    aria-expanded="false"
    class="sc-kLhKbu ikZkRM aje-btn aje-btn--primary is-auto"
    type="button"
  >
    Trigger
  </button>
</DocumentFragment>
`;

exports[`OverlayTrigger > Matches the open snapshot 1`] = `
<DocumentFragment>
  <button
    aria-controls="react-aria-:r1:"
    aria-expanded="true"
    class="sc-kLhKbu ikZkRM aje-btn aje-btn--primary is-auto"
    data-focus-visible="true"
    data-pressed="true"
    type="button"
  >
    Trigger
  </button>
</DocumentFragment>
`;
